### PR TITLE
223/favorite spaces

### LINKED ISSUE #222: Add “Recent Spaces”
Add as a Space Filter:
- Order by name
- Order by last added
- ...

### LINKED ISSUE #223: Add filter: “Favourites”

### PR DESCRIPTION
**Pull request by pyphilia**
closes #223
closes #222

### CODE REVIEW

**Comment by pyphilia** [comment]
I forgot the screenshots:
image
We have now Saved Spaces which was the usual Home page. All MediaCards have the star to set the corresponding space as favorite. It is in pretty yellow when it is on.

image
In Home we have now Spaces and Favorite Spaces. Recent Spaces are spaces we visit (when saved) or load. We save and display the five most recent spaces.

File: src/Home.js [hunk 1]
```diff
@@ -84,26 +186,41 @@ class Home extends Component {
...
     return (
-      <Main id={HOME_MAIN_ID} showSearch>
-        <SpaceGrid spaces={filteredSpaces} showActions saved />
+      <Main showSearch handleOnSearch={this.handleOnSearch} id={HOME_MAIN_ID}>
+        {noSpacesAvailable ? (
+          <NoSpacesAvailable />
+        ) : (
+          <div className={classes.home}>
+            {this.renderRecentSpaces()}
+            <br />
```
Comment (juancarlosfarah) [inline src/Home.js:197]: Maybe add a `Divider` component here?

Comment (juancarlosfarah) [reply src/Home.js:197]: From Material UI

File: src/components/space/FavoriteButton.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    });
+  };
+
+  render() {
+    const { classes, t, favoriteSpaces, spaceId } = this.props;
+
+    const isFavorite = favoriteSpaces.includes(spaceId);
+
+    return (
+      <Tooltip title={t('Add this Space to your Favorite Spaces')}>
```
Comment (juancarlosfarah) [inline src/components/space/FavoriteButton.js:57]: How about if you change the title if it's already a favorite?

File: src/components/space/SpaceScreen.js [hunk 3]
```diff
@@ -106,6 +118,17 @@ class SpaceScreen extends Component {
       // eslint-disable-next-line react/no-did-update-set-state
       this.setState({ selected: -1 });
     }
+
+    if (space && !space.isEmpty()) {
+      const spaceId = space.get('id');
+      const recentIndex = recentSpaces.indexOf(spaceId);
+
+      // update recent spaces if space is not recent
+      // or space is not most recent space
```
Comment (juancarlosfarah) [inline src/components/space/SpaceScreen.js:127]: How are you referring to the most recent space here? Wouldn't it be `0`?

File: src/langs/en.json [hunk 4]
```diff
@@ -59,6 +59,8 @@
     "You are currently synchronizing a space. Either accept or cancel the synchronization. The synchronization is definitive and all user input will be deleted.": "You are currently synchronizing a space. Either accept or cancel the synchronization. The synchronization is definitive and all user input will be deleted.",
     "Local Space Image": "Local Space Image",
     "Remote Space Image": "Remote Space Image",
-    "No Item to Display": "No Item to Display"
+    "No Item to Display": "No Item to Display",
+    "Add this Space to your Favorite Spaces": "Add this Space to your Favorite Spaces",
```
Comment (juancarlosfarah) [inline src/langs/en.json:63]: "Add this *space*" in small caps

File: src/langs/fr.json [hunk 5]
```diff
@@ -59,6 +59,8 @@
     "You are currently synchronizing a space. Either accept or cancel the synchronization. The synchronization is definitive and all user input will be deleted.": "Synchronisation d'un Espace en cours. La synchronisation peut être acceptée ou annulée. Si acceptée, la synchronisation est définitive et toute entrée utilisateur sera supprimée.",
     "Local Space Image": "Image Espace Locale",
     "Remote Space Image": "Image Espace En Ligne",
-    "No Item to Display": "Aucun Element à Afficher"
+    "No Item to Display": "Aucun Element à Afficher",
+    "Add this Space to your Favorite Spaces": "Add this Space to your Favorite Spaces",
```
Comment (juancarlosfarah) [inline src/langs/fr.json:63]: French?

File: src/components/space/SpaceScreen.js (hunk 3 above)
Comment (juancarlosfarah) [reply src/components/space/SpaceScreen.js:127]: Ah... I see. You're doing this in reverse.

File: src/reducers/authenticationReducer.js [hunk 6]
```diff
@@ -50,6 +90,9 @@ export const DEFAULT_USER_SETTINGS = {
 export const DEFAULT_USER = {
   geolocation: Map(),
   settings: { ...DEFAULT_USER_SETTINGS },
+  // redux doesn't seem to detect updates with sets
```
Comment (juancarlosfarah) [inline src/reducers/authenticationReducer.js:93]: That's weird. 🤔

File: src/reducers/authenticationReducer.js [hunk 7]
```diff
@@ -82,16 +126,16 @@ export default (state = INITIAL_STATE, { type, payload }) => {
       );
     case SIGN_IN_SUCCEEDED: {
       return state
-        .setIn(['user'], { ...DEFAULT_USER, ...payload })
+        .setIn(['user'], fromJS({ ...DEFAULT_USER, ...payload }))
```
Comment (juancarlosfarah) [inline src/reducers/authenticationReducer.js:129]: Why the change to `fromJS`?

File: src/reducers/authenticationReducer.js [hunk 8]
```diff
@@ -112,6 +156,16 @@ export default (state = INITIAL_STATE, { type, payload }) => {
     case SET_USER_MODE_SUCCEEDED:
     case GET_USER_MODE_SUCCEEDED:
       return state.setIn(['user', 'settings', 'userMode'], payload);
+    case SET_SPACE_AS_FAVORITE_SUCCEEDED:
```
Comment (juancarlosfarah) [inline src/reducers/authenticationReducer.js:159]: This could be simplified if you separate into two (add/remove), but if you're comfortable with this way, then no problem. 👍

File: src/components/space/SpaceScreen.js (hunk 3 above)
Comment (pyphilia) [reply src/components/space/SpaceScreen.js:127]: Yes it was easier to add spaces at the end of the List and remove the first element if the max number was reached.

File: src/reducers/authenticationReducer.js (hunk 6 above)
Comment (pyphilia) [reply src/reducers/authenticationReducer.js:93]: I really tried but nothing seemed to update.. maybe we try later.

File: src/reducers/authenticationReducer.js (hunk 7 above)
Comment (pyphilia) [reply src/reducers/authenticationReducer.js:129]: To keep the object as immutable. Plus `fromJS` deeply transforms the properties to immutable objects.

File: src/Home.js (hunk 1 above)
Comment (pyphilia) [reply src/Home.js:197]: I changed to it

image
